File: android/lib/src/main/kotlin/com/mia21/services/StreamingService.kt
/**
 * StreamingService.kt
 * Mia21
 *
 * Created on November 25, 2025.
 * Copyright © 2025 Mia21. All rights reserved.
 *
 * Description:
 * Service layer for streaming chat operations.
 * Handles text and voice streaming with SSE event parsing.
 */

package com.mia21.services

import com.mia21.models.*
import com.mia21.network.APIClient
import com.mia21.network.APIEndpoint
import com.mia21.network.HTTPMethod
import com.mia21.utils.Logger
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.filterIsInstance
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import org.json.JSONObject
import android.util.Base64
import kotlinx.coroutines.channels.awaitClose

/**
 * Service for streaming operations
 */
class StreamingService(private val apiClient: APIClient) {
    
    /**
     * Stream chat messages (text only)
     * This is a convenience method that uses streamChatWithVoice internally
     * with voiceConfig = null, extracting only text events.
     */
    fun streamChat(
        userId: String,
        messages: List<ChatMessage>,
        options: ChatOptions,
        customerLlmKey: String?,
        currentSpace: String?
    ): Flow<String> {
        return streamChatWithVoice(
            userId = userId,
            messages = messages,
            options = options,
            voiceConfig = null,
            customerLlmKey = customerLlmKey,
            currentSpace = currentSpace
        )
            .filterIsInstance<StreamEvent.Text>()
            .map { it.content }
    }
    
    /**
     * Stream chat messages with voice synthesis support
     * @param userId User identifier
     * @param messages List of conversation history messages
     * @param options Chat configuration options
     * @param voiceConfig Voice synthesis configuration (null for text-only)
     * @param customerLlmKey Optional customer LLM key
     * @param currentSpace Current active space ID
     * @return Flow of StreamEvent (text chunks, audio chunks, completion, errors)
     */
    fun streamChatWithVoice(
        userId: String,
        messages: List<ChatMessage>,
        options: ChatOptions,
        voiceConfig: VoiceConfig?,
        customerLlmKey: String?,
        currentSpace: String?
    ): Flow<StreamEvent> = callbackFlow {
        if (currentSpace == null && options.spaceId == null) {
            close(Mia21Exception.ChatNotInitializedException())
            return@callbackFlow
        }
        
        val messagesList = messages.map { mapOf("role" to it.role.name.lowercase(), "content" to it.content) }
        
        val body = mutableMapOf<String, Any?>(
            "user_id" to userId,
            "space_id" to (options.spaceId ?: currentSpace ?: "default_space"),
            "messages" to messagesList,
            "llm_type" to (options.llmType ?: LLMType.OPENAI).value
        )
        
        if (voiceConfig != null && voiceConfig.enabled) {
            body["response_mode"] = "stream_voice"
            body["voice_config"] = buildVoiceConfig(voiceConfig)
        } else {
            body["response_mode"] = "stream_text"
        }
        
        options.temperature?.let { body["temperature"] = it }
        options.maxTokens?.let { body["max_tokens"] = it }
        options.botId?.let { body["bot_id"] = it }
        options.conversationId?.let { body["conversation_id"] = it }
        options.voiceId?.let { body["voice_id"] = it }
        (options.customerLlmKey ?: customerLlmKey)?.let { body["customer_llm_key"] = it }
        
        val endpoint = APIEndpoint(
            path = "/chat/stream",
            method = HTTPMethod.POST,
            body = body
        )
        
        var textChunkCount = 0
        var audioChunkCount = 0
        
        val job = launch {
            try {
                apiClient.performStreamRequest(endpoint).collect { data ->
                    processStreamData(data) { event ->
                        when (event) {
                            is StreamEvent.Text -> {
                                textChunkCount++
                                Logger.debug("Text Chunk #$textChunkCount: ${event.content}")
                                trySend(event)
                            }
                            is StreamEvent.Audio -> {
                                audioChunkCount++
                                Logger.debug("Audio Chunk #$audioChunkCount: ${event.audioData.size} bytes")
                                trySend(event)
                            }
                            is StreamEvent.TextComplete -> {
                                trySend(event)
                            }
                            is StreamEvent.Done -> {
                                Logger.info("Stream completed. Text: $textChunkCount, Audio: $audioChunkCount")
                                trySend(event)
                                close()
                            }
                            is StreamEvent.Error -> {
                                Logger.error("Stream error: ${event.exception.message}")
                                trySend(event)
                                close(event.exception)
                            }
                        }
                    }
                }
                
                trySend(StreamEvent.Done(null))
            } catch (e: Exception) {
                Logger.error("Stream failed: ${e.message}")
                trySend(StreamEvent.Error(e))
                close(e)
            }
        }
        
        awaitClose {
            job.cancel()
        }
    }
    
    /**
     * Stream completion using the OpenAI-compatible endpoint.
     * No bot/space pre-configuration required - include system message in the messages list.
     */
    fun streamComplete(
        userId: String,
        messages: List<ChatMessage>,
        options: CompletionOptions
    ): Flow<String> = callbackFlow {
        Logger.debug("Starting streaming completion with ${messages.size} messages")
        
        // Build OpenAI-compatible messages array
        val messagesList = messages.map { msg ->
            mapOf("role" to msg.role.name.lowercase(), "content" to msg.content)
        }
        
        val body = mutableMapOf<String, Any?>(
            "model" to options.model,
            "messages" to messagesList,
            "stream" to true
        )
        
        options.temperature?.let { body["temperature"] = it }
        options.maxTokens?.let { body["max_tokens"] = it }
        
        // Build headers for OpenAI-compatible endpoint
        val headers = mutableMapOf("X-User-Id" to userId)
        options.spaceId?.let { headers["X-Space-Id"] = it }
        options.botId?.let { headers["X-Bot-Id"] = it }
        
        val endpoint = APIEndpoint(
            path = "/v1/chat/completions",
            method = HTTPMethod.POST,
            body = body,
            headers = headers
        )
        
        val job = launch {
            try {
                apiClient.performStreamRequest(endpoint).collect { data ->
                    processOpenAIStreamData(data) { content ->
                        trySend(content)
                    }
                }
            } catch (e: Exception) {
                Logger.error("Streaming completion failed: ${e.message}")
                close(e)
            }
        }
        
        awaitClose {
            job.cancel()
        }
    }
    
    /**
     * Process OpenAI-style streaming data
     */
    private fun processOpenAIStreamData(data: String, onChunk: (String) -> Unit) {
        val trimmed = data.trim()
        
        if (trimmed.isEmpty() || trimmed == "[DONE]") {
            return
        }
        
        try {
            val json = JSONObject(trimmed)
            val choices = json.optJSONArray("choices") ?: return
            if (choices.length() == 0) return
            
            val firstChoice = choices.getJSONObject(0)
            val delta = firstChoice.optJSONObject("delta") ?: return
            val content = delta.optString("content", "")
            
            if (content.isNotEmpty()) {
                onChunk(content)
            }
        } catch (e: Exception) {
            // Not valid OpenAI format, skip
        }
    }
    
    /**
     * Build voice configuration dictionary for API request
     */
    private fun buildVoiceConfig(config: VoiceConfig): Map<String, Any?> {
        val voiceDict = mutableMapOf<String, Any?>(
            "enabled" to true
        )
        
        config.voiceId?.let { voiceDict["voice_id"] = it }
        config.elevenlabsApiKey?.let { voiceDict["elevenlabs_api_key"] = it }
        config.stability?.let { voiceDict["stability"] = it }
        config.similarityBoost?.let { voiceDict["similarity_boost"] = it }
        
        return voiceDict
    }
    
    /**
     * Process stream data (performStreamRequest already extracts data from SSE)
     */
    private fun processStreamData(data: String, onEvent: (StreamEvent) -> Unit) {
        val trimmedForCheck = data.trim()
        
        if (trimmedForCheck.isEmpty()) {
            return
        }
        
        if (trimmedForCheck.equals("[DONE]", ignoreCase = true)) {
            Logger.info("Received [DONE] marker")
            return
        }
        
        try {
            val json = JSONObject(data.trim())
            
            if (json.has("content")) {
                val textContent = json.getString("content")
                if (textContent.isNotEmpty()) {
                    onEvent(StreamEvent.Text(textContent))
                }
            }
            
            if (json.has("audio")) {
                val audioBase64 = json.getString("audio")
                try {
                    val audioData = Base64.decode(audioBase64, Base64.DEFAULT)
                    onEvent(StreamEvent.Audio(audioData))
                } catch (e: IllegalArgumentException) {
                    Logger.error("Failed to decode audio Base64: ${e.message}")
                }
            }
            
            return
        } catch (e: Exception) {
            val textContent = extractTextContent(data)
            if (textContent != null) {
                onEvent(StreamEvent.Text(textContent))
            }
        }
    }
    
    /**
     * Extract text content from data string, filtering out structured data
     * Preserves all spaces in the original content
     */
    private fun extractTextContent(content: String): String? {
        val trimmedForCheck = content.trim()
        
        if (trimmedForCheck.isEmpty() || trimmedForCheck.equals("[DONE]", ignoreCase = true)) {
            return null
        }
        
        if (trimmedForCheck.startsWith("{") || trimmedForCheck.startsWith("[")) {
            try {
                val json = JSONObject(trimmedForCheck)
                if (json.has("content")) {
                    val textContent = json.getString("content")
                    if (textContent.isNotEmpty()) {
                        return textContent
                    }
                }
                return null
            } catch (e: Exception) {
                // Not valid JSON, continue to check for function calls
            }
        }
        
        if (trimmedForCheck.contains("'type': 'function_call'") ||
            trimmedForCheck.contains("\"type\": \"function_call\"") ||
            trimmedForCheck.contains("'function_call'") ||
            trimmedForCheck.contains("\"function_call\"")) {
            return null
        }
        
        return content
    }
}

File: ios/Sources/Mia21/Services/StreamingService.swift
//
//  StreamingService.swift
//  Mia21
//
//  Created on November 4, 2025.
//  Copyright © 2025 Mia21. All rights reserved.
//
//  Description:
//  Service layer for streaming operations.
//  Handles text and voice streaming with SSE event parsing.
//

import Foundation

// MARK: - Streaming Service Protocol

protocol StreamingServiceProtocol {
  func streamChat(
    userId: String,
    messages: [ChatMessage],
    options: ChatOptions,
    customerLlmKey: String?,
    currentSpace: String?,
    onChunk: @escaping (String) -> Void
  ) async throws

  func streamChatWithVoice(
    userId: String,
    messages: [ChatMessage],
    options: ChatOptions,
    voiceConfig: VoiceConfig?,
    customerLlmKey: String?,
    currentSpace: String?,
    onEvent: @escaping (StreamEvent) -> Void
  ) async throws
  
  func streamComplete(
    userId: String,
    messages: [ChatMessage],
    options: CompletionOptions,
    onChunk: @escaping (String) -> Void
  ) async throws
}

// MARK: - Streaming Service Implementation

final class StreamingService: StreamingServiceProtocol {

  // MARK: - Properties

  private let apiClient: APIClientProtocol

  // MARK: - Initialization

  init(apiClient: APIClientProtocol) {
    self.apiClient = apiClient
  }

  // MARK: - Public Methods

  func streamChat(
    userId: String,
    messages: [ChatMessage],
    options: ChatOptions,
    customerLlmKey: String?,
    currentSpace: String?,
    onChunk: @escaping (String) -> Void
  ) async throws {
    guard currentSpace != nil || options.spaceId != nil else {
      throw Mia21Error.chatNotInitialized
    }

    let messagesDictArray = messages.map { ["role": $0.role.rawValue, "content": $0.content] }

    var body: [String: Any] = [
      "user_id": userId,
      "space_id": options.spaceId ?? currentSpace ?? "default_space",
      "messages": messagesDictArray,
      "llm_type": (options.llmType ?? .openai).rawValue,
      "stream": true
    ]

    if let conversationId = options.conversationId {
      body["conversation_id"] = conversationId
    }

    addCommonOptions(to: &body, options: options, customerLlmKey: customerLlmKey)

    let endpoint = APIEndpoint(path: "/chat/stream", method: .post, body: body)
    let stream = try await apiClient.performStreamRequest(endpoint)

    for try await data in stream {
      if let line = String(data: data, encoding: .utf8) {
        processTextStreamLine(line, onChunk: onChunk)
      }
    }
  }

  func streamChatWithVoice(
    userId: String,
    messages: [ChatMessage],
    options: ChatOptions,
    voiceConfig: VoiceConfig?,
    customerLlmKey: String?,
    currentSpace: String?,
    onEvent: @escaping (StreamEvent) -> Void
  ) async throws {
    guard currentSpace != nil || options.spaceId != nil else {
      throw Mia21Error.chatNotInitialized
    }

    let messagesDictArray = messages.map { ["role": $0.role.rawValue, "content": $0.content] }

    var body: [String: Any] = [
      "user_id": userId,
      "space_id": options.spaceId ?? currentSpace ?? "default_space",
      "messages": messagesDictArray,
      "llm_type": (options.llmType ?? .openai).rawValue
    ]

    // Configure response mode based on voice settings
    if let voiceConfig = voiceConfig, voiceConfig.enabled {
      body["response_mode"] = "stream_voice"
      body["voice_config"] = buildVoiceConfig(voiceConfig)
    } else {
      body["response_mode"] = "stream_text"
    }

    if let conversationId = options.conversationId {
      body["conversation_id"] = conversationId
    }

    addCommonOptions(to: &body, options: options, customerLlmKey: customerLlmKey)

    let endpoint = APIEndpoint(path: "/chat/stream", method: .post, body: body)
    let stream = try await apiClient.performStreamRequest(endpoint)

    var parser = SSEParser()

    for try await data in stream {
      if let line = String(data: data, encoding: .utf8) {
        parser.parse(line: line) { event in
          onEvent(event)
        }
      }
    }

    // Send final done event
    onEvent(.done(nil))
  }
  
  // MARK: - OpenAI-Compatible Streaming Completions
  
  func streamComplete(
    userId: String,
    messages: [ChatMessage],
    options: CompletionOptions,
    onChunk: @escaping (String) -> Void
  ) async throws {
    logInfo("Starting streaming completion with \(messages.count) messages")
    
    // Build OpenAI-compatible messages array
    let messagesArray = messages.map { msg -> [String: String] in
      return ["role": msg.role.rawValue, "content": msg.content]
    }
    
    var body: [String: Any] = [
      "model": options.model,
      "messages": messagesArray,
      "stream": true
    ]
    
    if let temperature = options.temperature {
      body["temperature"] = temperature
    }
    if let maxTokens = options.maxTokens {
      body["max_tokens"] = maxTokens
    }
    
    // Build headers for OpenAI-compatible endpoint
    var headers: [String: String] = [
      "X-User-Id": userId
    ]
    if let spaceId = options.spaceId {
      headers["X-Space-Id"] = spaceId
    }
    if let botId = options.botId {
      headers["X-Bot-Id"] = botId
    }
    
    let endpoint = APIEndpoint(path: "/v1/chat/completions", method: .post, body: body, headers: headers)
    let stream = try await apiClient.performStreamRequest(endpoint)
    
    for try await data in stream {
      if let line = String(data: data, encoding: .utf8) {
        processOpenAIStreamLine(line, onChunk: onChunk)
      }
    }
  }

  // MARK: - Private Methods
  
  /// Simplified: Accept only text/audio content, filter out all objects/structured data
  /// Returns the text content to display, or nil if it should be skipped
  static func extractTextContent(from content: String) -> String? {
    // Skip truly empty content or [DONE] marker
    // Note: We keep whitespace-only content (like " ") as it may be valid text between words
    if content.isEmpty || content == "[DONE]" {
      return nil
    }
    
    let trimmed = content.trimmingCharacters(in: .whitespaces)
    
    // Skip [DONE] marker even with surrounding whitespace
    if trimmed == "[DONE]" {
      return nil
    }
    
    // If it's JSON, only extract "content" field (actual text)
    if trimmed.hasPrefix("{") || trimmed.hasPrefix("[") {
      if let data = trimmed.data(using: .utf8),
         let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
         let textContent = json["content"] as? String, !textContent.isEmpty {
        return textContent
      }
      // All other JSON (function calls, objects, etc.) - skip
      return nil
    }
    
    // If it contains function call patterns (Python dict or JSON style) - skip
    if trimmed.contains("'type': 'function_call'") ||
       trimmed.contains("\"type\": \"function_call\"") ||
       trimmed.contains("'function_call'") ||
       trimmed.contains("\"function_call\"") {
      return nil
    }
    
    // Otherwise, it's plain text - accept it
    return content
  }

  private func addCommonOptions(to body: inout [String: Any], options: ChatOptions, customerLlmKey: String?) {
    if let temperature = options.temperature {
      body["temperature"] = temperature
    }
    if let maxTokens = options.maxTokens {
      body["max_tokens"] = maxTokens
    }
    if let botId = options.botId {
      body["bot_id"] = botId
    }
    if let voiceId = options.voiceId {
      body["voice_id"] = voiceId
    }

    let llmKey = options.customerLlmKey ?? customerLlmKey
    if let llmKey = llmKey {
      body["customer_llm_key"] = llmKey
    }

    if let spaceConfig = options.spaceConfig {
      let encoder = JSONEncoder()
      encoder.keyEncodingStrategy = .convertToSnakeCase
      if let configData = try? encoder.encode(spaceConfig),
         let configDict = try? JSONSerialization.jsonObject(with: configData) as? [String: Any] {
        body["space_config"] = configDict
      }
    }
  }

  private func buildVoiceConfig(_ config: VoiceConfig) -> [String: Any] {
    var voiceDict: [String: Any] = ["enabled": true]

    if let voiceId = config.voiceId {
      voiceDict["voice_id"] = voiceId
    }
    if let apiKey = config.elevenlabsApiKey {
      voiceDict["elevenlabs_api_key"] = apiKey
    }
    if let stability = config.stability {
      voiceDict["stability"] = stability
    }
    if let similarityBoost = config.similarityBoost {
      voiceDict["similarity_boost"] = similarityBoost
    }

    return voiceDict
  }

  private func processTextStreamLine(_ line: String, onChunk: @escaping (String) -> Void) {
    // Only trim to check if empty, but preserve original content with newlines
    let trimmedForCheck = line.trimmingCharacters(in: .whitespaces)

    // Skip completely empty lines (blank lines between SSE events)
    if trimmedForCheck.isEmpty {
      return
    }

    // Check if line has "data: " prefix (SSE format)
    if line.hasPrefix("data: ") {
      let content = String(line.dropFirst(6)) // Remove "data: " prefix
      
      // Empty data line means a newline in the content
      // (SSE sends "data: " with no content to represent line breaks)
      if content.isEmpty {
        onChunk("\n")
        return
      }
      
      // Use general text extraction - only show actual messages
      if let textContent = StreamingService.extractTextContent(from: content) {
        onChunk(textContent)
      }
      // If extractTextContent returns nil, it's structured data - skip it silently
    } else if trimmedForCheck != "[DONE]" {
      // If no prefix, check if it's actual text content
      if let textContent = StreamingService.extractTextContent(from: line) {
        onChunk(textContent)
      }
    }
  }
  
  private func processOpenAIStreamLine(_ line: String, onChunk: @escaping (String) -> Void) {
    let trimmed = line.trimmingCharacters(in: .whitespaces)
    
    // Skip empty lines
    if trimmed.isEmpty {
      return
    }
    
    // Check for SSE data prefix
    guard line.hasPrefix("data: ") else {
      return
    }
    
    let content = String(line.dropFirst(6))
    
    // Check for [DONE] marker
    if content == "[DONE]" {
      return
    }
    
    // Parse OpenAI streaming format
    guard let data = content.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let choices = json["choices"] as? [[String: Any]],
          let firstChoice = choices.first,
          let delta = firstChoice["delta"] as? [String: Any],
          let textContent = delta["content"] as? String else {
      return
    }
    
    onChunk(textContent)
  }
}

// MARK: - SSE Parser

private struct SSEParser {
  private var currentEvent: String?
  private var chunkCount = 0
  private var audioChunkCount = 0

  mutating func parse(line: String, onEvent: (StreamEvent) -> Void) {
    // Only trim to check if truly empty, preserve original for content extraction
    let trimmedForCheck = line.trimmingCharacters(in: .whitespaces)

    if trimmedForCheck.isEmpty {
      return
    }

    // Parse event type - use original line to preserve spacing
    if line.hasPrefix("event: ") {
      currentEvent = String(line.dropFirst(7)).trimmingCharacters(in: .whitespaces)
      logDebug("Event type: \(currentEvent ?? "nil")")
      return
    }

    // Parse data - use original line to preserve content with spaces
    if line.hasPrefix("data: ") {
      let dataString = String(line.dropFirst(6))

      // Check for [DONE] marker
      if dataString == "[DONE]" {
        logInfo("Stream completed. Text: \(chunkCount), Audio: \(audioChunkCount)")
        return
      }
      
      // Empty data line means a newline in the content
      if dataString.isEmpty {
        chunkCount += 1
        onEvent(.text("\n"))
        return
      }

      // Simplified: Only accept text/audio content, filter out all objects
      
      // Try to parse as JSON
      if let data = dataString.data(using: .utf8),
         let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
        
        // Only handle "content" (text) or "audio" fields - skip everything else
        if let textContent = json["content"] as? String, !textContent.isEmpty {
          chunkCount += 1
          logDebug("Text Chunk #\(chunkCount): \(textContent)")
          onEvent(.text(textContent))
        }
        
        if let audioBase64 = json["audio"] as? String,
           let audioData = Data(base64Encoded: audioBase64) {
          audioChunkCount += 1
          logDebug("Audio Chunk #\(audioChunkCount): \(audioData.count) bytes")
          onEvent(.audio(audioData))
        }
        
        // All other JSON fields (function_call, tool_call, error, etc.) - skip silently
        return
      }
      
      // Not JSON - use text extraction to filter structured data
      if let textContent = StreamingService.extractTextContent(from: dataString) {
        switch currentEvent {
        case "text_complete":
          onEvent(.textComplete)
        case "error":
          onEvent(.error(Mia21Error.streamingError(dataString)))
        default:
          chunkCount += 1
          onEvent(.text(textContent))
        }
      }
      // If extractTextContent returns nil, it's structured data - skip silently

      currentEvent = nil
    }
  }
}


